test/test_charm.py

```diff
@@ -1253,3 +1261,316 @@ def test_meta_charm_user_default():
+    @property
+    def config_class(self) -> type[_ConfigProtocol]:
+        return MyConfig
+
+
+# Note that we would really like to have kw_only=True here as well, but that's
+# not available in Python 3.8.
+@dataclasses.dataclass(frozen=True)
+class MyDataclassConfig:
+    my_bool: bool | None = None
+    my_int: int = 42
+    my_float: float = 3.14
+    my_str: str = 'foo'
+    my_secret: ops.Secret | None = None
+
+    def __post_init__(self):
+        if self.my_int < 0:
+            raise ValueError('my_int must be zero or positive')
+
+
+class MyDataclassCharm(BaseTestCharm):
+    @property
+    def config_class(self) -> type[_ConfigProtocol]:
+        return MyDataclassConfig
+
+
+_test_classes: list[type[ops.CharmBase]] = [MyCharm, MyDataclassCharm]
+
+if pydantic:
+
+    @pydantic.dataclasses.dataclass(frozen=True, config={'arbitrary_types_allowed': True})
+    class MyPydanticDataclassConfig:
+        my_bool: bool | None = pydantic.Field(None)
+        my_int: int = pydantic.Field(42)
+        my_float: float = pydantic.Field(3.14)
+        my_str: str = pydantic.Field('foo')
+        my_secret: ops.Secret | None = pydantic.Field(None)
+
+        @pydantic.field_validator('my_int')
+        @classmethod
+        def validate_my_int(cls, my_int: int) -> int:
+            if my_int < 0:
+                raise ValueError('my_int must be zero or positive')
+            return my_int
+
+    class MyPydanticDataclassCharm(BaseTestCharm):
+        @property
+        def config_class(self) -> type[_ConfigProtocol]:
+            return MyPydanticDataclassConfig
+
+    class MyPydanticBaseModelConfig(pydantic.BaseModel):
+        my_bool: bool | None = pydantic.Field(None)
+        my_int: int = pydantic.Field(42)
+        my_float: float = pydantic.Field(3.14)
+        my_str: str = pydantic.Field('foo')
+        my_secret: ops.Secret | None = pydantic.Field(None)
+
+        @pydantic.field_validator('my_int')
+        @classmethod
+        def validate_my_int(cls, my_int: int) -> int:
+            if my_int < 0:
+                raise ValueError('my_int must be zero or positive')
+            return my_int
+
+        class Config:
+            arbitrary_types_allowed = True
+            frozen = True
+
+    class MyPydanticBaseModelCharm(BaseTestCharm):
+        @property
+        def config_class(self) -> type[_ConfigProtocol]:
+            return MyPydanticBaseModelConfig
+
+    _test_classes.extend((MyPydanticDataclassCharm, MyPydanticBaseModelCharm))
+
+
+_config = """
+options:
+    my-bool:
+        type: boolean
+    my-int:
+        type: int
+        default: 42
+    my-float:
+        type: float
+        default: 3.14
+    my-str:
+        type: string
+        default: foo
+    my-secret:
+        type: secret
+"""
+
+
+@pytest.mark.parametrize('charm_class', _test_classes)
+def test_config_init(charm_class: type[ops.CharmBase], request: pytest.FixtureRequest):
+    harness = testing.Harness(charm_class, config=_config)
```

Comment:
@benhoyt I'll use Scenario for the new package. For this one, with tests in `test_charm`, would you still prefer that I rewrite these to use Scenario?

Reply:
No, this is fine then in `test_charm.py`